cli/cli.go

```diff
@@ -137,37 +137,53 @@ func (c ConfigVersion) IsValid() bool {
 	return c != 0 && c <= V2
 }
 
+// EndpointFlag defines the http(s) Enpoint
+type EndpointFlag string
+
+func (ef *EndpointFlag) String() string {
+	return string(*ef)
+}
+
+// Set validates Endpoint before setting it
+func (ef *EndpointFlag) Set(s string) error {
+	nurl, err := url.ParseRequestURI(s)
+	if err != nil {
+		return err
+	}
+	*ef = EndpointFlag(nurl.String())
+	return nil
+}
+
+// Type returns a string that uniquely represents this flag's type.
+func (ef *EndpointFlag) Type() string {
+	return "string"
+}
+
+// EndpointURL returns Parsed URL type of endpoint
+func (ef EndpointFlag) EndpointURL() *url.URL {
+	nurl, _ := url.ParseRequestURI(ef.String())
+	return nurl
+}
```

Comment:
Parsed endpoint URL is needed in a couple of places, [here](https://github.com/hasura/graphql-engine/blob/master/cli/cli.go#L156) and [here](https://github.com/hasura/graphql-engine/blob/master/cli/migrate/util.go#L134) thus I added this function. Since any errors while parsing endpoint will be caught in `Set()` I didn't catch errors here but I'm not sure if this is the best way to do it.

Reply:
Got it. In that case, we can define the endpoint flag as of type `url.URL` instead of `string`. Then, we will only need to have `Endpoint url.URL` [here](https://github.com/hasura/graphql-engine/blob/master/cli/cli.go#L141-L152) instead of having both `Endpoint` and `ParsedEndpoint`.

Wherever we need it as a string, we can use [`String()`](https://pkg.go.dev/net/url?tab=doc#URL.String) method of `url.URL`.

Reply:
I tried implementing that as well, but I wasn't sure [here](https://github.com/hasura/graphql-engine/blob/master/cli/cli.go#L530) how to convert type string to type `EndpointFlag`.

Reply:
I think `viper` should take care of validating the flag if we use `v.Get` and then execute a type assertion over the interface value that is returned. Something like `(Endpoint)(v.Get("endpoint))`